### DISCUSSION TITLE
Integrating With Github

### DISCUSSION THREAD

**Discussion by mohitprasanth** [WeblateOrg/weblate#6765] [General] [answered]
I am using weblate-docker and 
I went through https://docs.weblate.org/en/latest/vcs.html#vcs-repos

but not able to push my changes as PR to GITHUB.
it says Bad Credentials. OR Fork creation failed: Bad credentials , what ever below methods I try.

Basically I generated SSH Key in weblate and used the same for generating token on Github end.
I configured WEBLATE_GITHUB_USERNAME and WEBLATE_GITHUB_TOKEN.
For WEBLATE_GITHUB_TOKEN I tried 
 * Using SSH Token (Settings -> SSH and GPG key)
 * Using Deployment Key for that repo with write access. (Repo -> settings -> Deployment Key) 
 * Added Weblate Bot Integration and provided necessary repo permissions.

None of the above is working for me, it still gives Bad Credentials error.
I am the owner of the repo.

Can you please let me know if I am missing something here ?
Also if someone can point out the steps that worked for them to setup Github Integration (for private repo) that would be helpful.

[Update]
Adding Further Details and my debugging.

- Using docker-compose made my instance up.
- Logged in as admin
- Created a Project
- Generated SSH (Settings -> SSH Key -> Generate)
- In my GitHub , Create a Deployment Key with write access. Using Above SSH Key.
- Configured above generated token in my environment file.
WEBLATE_GITHUB_USERNAME=Weblate
WEBLATE_GITHUB_TOKEN=2rWMUchDXXXXXXXXXXXXXXX
- Restarted using docker-compose.
- Create a Component under the above project
with all git urls as : 
git@github.com:mohitprasanth/WEBLATE_CONTENT.git
- Added a new translation.
- Repository Maintainance -> Commit -> Push

I AM GETTING BELOW EXCEPTION:
**""Fork creation failed: Bad credentials""**

**My Debugging Info: (ON My Public Repo)**
`$ docker exec it CONTAINER-ID sh`

`$ pwd`
/app/data/vcs/cm-dev/cm-commons

`$ git status`
On branch main
Your branch is ahead of 'origin/main' by 1 commit.
  (use "git push" to publish your local commits)

`$ git push`
The authenticity of host 'github.com (13.234.176.102)' can't be established.
RSA key fingerprint is SHA256:nThbg6kXUpJWGl7E1IGOCspRomTxdCARLviKw6E5SY8.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added 'github.com,13.234.176.102' (RSA) to the list of known hosts.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

`$ echo $WEBLATE_GITHUB_USERNAME`
Weblate

`$ echo $WEBLATE_GITHUB_TOKEN`
2rWMUchDXXXXXXXXXXXXXXX

`$ eval "$(ssh-agent -s)"`
Agent pid 480

`$ ssh-add -l -E sha256`
The agent has no identities.

`$ ssh -vT git@github.com`
OpenSSH_8.4p1 Debian-5, OpenSSL 1.1.1k  25 Mar 2021
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug1: Connecting to github.com [13.234.210.38] port 22.
debug1: Connection established.
debug1: identity file /home/weblate/.ssh/id_rsa type -1
debug1: identity file /home/weblate/.ssh/id_rsa-cert type -1
debug1: identity file /home/weblate/.ssh/id_dsa type -1
debug1: identity file /home/weblate/.ssh/id_dsa-cert type -1
debug1: identity file /home/weblate/.ssh/id_ecdsa type -1
debug1: identity file /home/weblate/.ssh/id_ecdsa-cert type -1
debug1: identity file /home/weblate/.ssh/id_ecdsa_sk type -1
debug1: identity file /home/weblate/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /home/weblate/.ssh/id_ed25519 type -1
debug1: identity file /home/weblate/.ssh/id_ed25519-cert type -1
debug1: identity file /home/weblate/.ssh/id_ed25519_sk type -1
debug1: identity file /home/weblate/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /home/weblate/.ssh/id_xmss type -1
debug1: identity file /home/weblate/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.4p1 Debian-5
debug1: Remote protocol version 2.0, remote software version babeld-2aa5ef1f
debug1: no match: babeld-2aa5ef1f
debug1: Authenticating to github.com:22 as 'git'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: rsa-sha2-512
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ssh-rsa SHA256:nThbg6kXUpJWGl7E1XXXXXXXXX
debug1: Host 'github.com' is known and matches the RSA host key.
debug1: Found key in /home/weblate/.ssh/known_hosts:1
Warning: Permanently added the RSA host key for IP address '13.234.210.38' to the list of known hosts.
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /home/weblate/.ssh/id_rsa
debug1: Will attempt key: /home/weblate/.ssh/id_dsa
debug1: Will attempt key: /home/weblate/.ssh/id_ecdsa
debug1: Will attempt key: /home/weblate/.ssh/id_ecdsa_sk
debug1: Will attempt key: /home/weblate/.ssh/id_ed25519
debug1: Will attempt key: /home/weblate/.ssh/id_ed25519_sk
debug1: Will attempt key: /home/weblate/.ssh/id_xmss
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ssh-dss-cert-v01@openssh.com,sk-ssh-ed25519@openssh.com,sk-ecdsa-sha2-nistp256@openssh.com,ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa,ssh-dss>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Trying private key: /home/weblate/.ssh/id_rsa
debug1: Trying private key: /home/weblate/.ssh/id_dsa
debug1: Trying private key: /home/weblate/.ssh/id_ecdsa
debug1: Trying private key: /home/weblate/.ssh/id_ecdsa_sk
debug1: Trying private key: /home/weblate/.ssh/id_ed25519
debug1: Trying private key: /home/weblate/.ssh/id_ed25519_sk
debug1: Trying private key: /home/weblate/.ssh/id_xmss
debug1: No more authentication methods to try.
git@github.com: Permission denied (publickey).

WHEN I ADD THIS:
`$ ssh-add /app/data/ssh/id_rsa`
Identity added: /app/data/ssh/id_rsa (Weblate)

I AM ABLE TO PUSH FROM TERMINAL.
BUT THE SAME ISSUE PERSISTS IN WEBLATE UI.

**Comment by nijel** [comment, marked as answer]
Weblate pushes as well, but fails to create the pull request.

WEBLATE_GITHUB_USERNAME should be username for which the WEBLATE_GITHUB_TOKEN is generated, so it is most likely wrong for your case.

**Reply by mohitprasanth** [reply]
Thanks for your reply @nijel  and the user name entry didnt effect the push flow.
I figured out the issue debugging more, I used the DEPLOYMENT_KEY as token, thats the reason its not working for me.

Finally these steps worked for me:

- Generated SSH Key via weblate UI.
- Added this Key in GitHub (in Deployment key section with write access)

Thus I am able to push my changes.

- Generated a token from GITHUB.
- Configured that TOKEN in WEBLATE_GITHUB_TOKEN in weblate environment variables.

Thus I am able to create a PR.

This worked for me.

We want to keep option of creating a new user `weblate` as last option. 
We are looking for solutions using Github apps or Github Api keys inorder to integrate with weblate.

I explored `Weblate Github App` but didnt found any documentation around it apart from [#1688](https://github.com/WeblateOrg/weblate/issues/1688)
Can you point me to documentation if its available somewhere ?

Also can you please let me know if there is a better way to do integration with GitHub ?
I want to  setup this in prod thus looking out for a better solution.

**Reply by nijel** [reply]
The user is used to create forks of projects to open pull requests from. It is the user for which you have generated TOKEN.

The GitHub app integration is not yet implemented. Right now, it is only used to trigger push notifications to Hosted Weblate. Still, you would have to register own app on GitHub marketplace for your Weblate instance.

**Reply by mohitprasanth** [reply]
Understood, Thanks for explaining @nijel .
Can you please let me know what is the ideal way to implement the GITHUB integration with weblate for prod like setup ?

>Still, you would have to register own app on GitHub marketplace for your Weblate instance.

can you please add more details about this ?

**Reply by nijel** [reply]
For now, the recommended approach is to have a dedicated user. See https://docs.weblate.org/en/latest/vcs.html#github-repositories

**Reply by mohitprasanth** [reply]
Thanks for your input @nijel
